File: src/lib/pdfReport.ts
import { isAllowedMediaHost } from "./kobo.js";

const KOBO2PDF_URL = "https://kobo2pdf.imtools.info";

export interface PdfReportConfig {
  template?: string;
  formTitle?: string;
}

function arrayBufferToBase64(buffer: ArrayBuffer): string {
  let binary = "";
  const bytes = new Uint8Array(buffer);
  for (let i = 0; i < bytes.byteLength; i++) {
    binary += String.fromCharCode(bytes[i]);
  }
  return btoa(binary);
}

export async function renderPdf(
  cfg: PdfReportConfig,
  submission: Record<string, unknown>,
  koboServer: string,
  koboToken: string
): Promise<{ ok: boolean; error?: string; pdfBytes?: ArrayBuffer }> {
  // ── 1. Fetch image attachments to embed in the PDF ───────────────────────
  const pdfAttachments: Array<{ filename: string; content_type: string; content: string }> = [];
  const rawAttachments = submission._attachments;
  if (Array.isArray(rawAttachments)) {
    for (const att of rawAttachments as Array<Record<string, unknown>>) {
      if (att.is_deleted) continue;
      const mimetype = String(att.mimetype ?? "");
      if (!mimetype.startsWith("image/")) continue;
      const downloadUrl = String(att.download_url ?? "");
      if (!downloadUrl || !isAllowedMediaHost(downloadUrl, koboServer)) continue;
      try {
        const mediaRes = await fetch(downloadUrl, {
          headers: { Authorization: `Token ${koboToken}` },
        });
        if (!mediaRes.ok) {
          console.error(`[pdf] Failed to fetch attachment ${att.media_file_basename}: ${mediaRes.status}`);
          continue;
        }
        const buf = await mediaRes.arrayBuffer();
        pdfAttachments.push({
          filename: String(att.media_file_basename ?? att.filename ?? ""),
          content_type: mimetype,
          content: arrayBufferToBase64(buf),
        });
      } catch (e) {
        console.error(`[pdf] Error fetching attachment ${att.media_file_basename}: ${e}`);
      }
    }
  }

  // ── 2. Call the kobo2pdf render service ─────────────────────────────────
  let pdfBytes: ArrayBuffer;
  try {
    const renderRes = await fetch(`${KOBO2PDF_URL}/render`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({
        template: cfg.template ?? "submission",
        data: submission,
        meta: {
          ...(cfg.formTitle ? { formTitle: cfg.formTitle } : {}),
          reportDate: new Date().toISOString().slice(0, 10),
        },
        ...(pdfAttachments.length ? { attachments: pdfAttachments } : {}),
      }),
    });
    if (!renderRes.ok) {
      const text = await renderRes.text().catch(() => "");
      return { ok: false, error: `kobo2pdf error ${renderRes.status}: ${text.slice(0, 200)}` };
    }
    const contentType = renderRes.headers.get("content-type") ?? "";
    if (!contentType.includes("application/pdf")) {
      return { ok: false, error: `kobo2pdf returned unexpected content-type: ${contentType}` };
    }
    pdfBytes = await renderRes.arrayBuffer();
  } catch (e) {
    return { ok: false, error: `Failed to reach kobo2pdf service: ${e}` };
  }

  return { ok: true, pdfBytes };
}

File: app/lib/pdf_report.py
"""
PDF generation via kobo2pdf.imtools.info.
Port of src/lib/pdfReport.ts.
"""

import base64
import logging
from datetime import date

import httpx

from .kobo import is_allowed_media_host

logger = logging.getLogger(__name__)

KOBO2PDF_URL = 'https://kobo2pdf.imtools.info'


def render_pdf(
    cfg: dict,
    submission: dict,
    kobo_server: str,
    kobo_token: str,
) -> dict:
    """
    Fetches image attachments and calls kobo2pdf to render a PDF.
    Returns {'ok': bool, 'error': str?, 'pdf_bytes': bytes?}.
    """
    # 1. Fetch image attachments to embed
    pdf_attachments = []
    for att in (submission.get('_attachments') or []):
        if att.get('is_deleted'):
            continue
        mimetype = str(att.get('mimetype') or '')
        if not mimetype.startswith('image/'):
            continue
        download_url = str(att.get('download_url') or '')
        if not download_url or not is_allowed_media_host(download_url, kobo_server):
            continue
        try:
            media_resp = httpx.get(
                download_url,
                headers={'Authorization': f'Token {kobo_token}'},
                timeout=30,
            )
            if not media_resp.is_success:
                logger.error('[pdf] Failed to fetch attachment %s: %s', att.get('media_file_basename'), media_resp.status_code)
                continue
            pdf_attachments.append({
                'filename': str(att.get('media_file_basename') or att.get('filename') or ''),
                'content_type': mimetype,
                'content': base64.b64encode(media_resp.content).decode('ascii'),
            })
        except Exception as exc:
            logger.error('[pdf] Error fetching attachment %s: %s', att.get('media_file_basename'), exc)

    # 2. Call the kobo2pdf render service
    try:
        body: dict = {
            'template': cfg.get('template') or 'submission',
            'data': submission,
            'meta': {
                'reportDate': date.today().isoformat(),
                **({'formTitle': cfg['formTitle']} if cfg.get('formTitle') else {}),
            },
        }
        if pdf_attachments:
            body['attachments'] = pdf_attachments

        render_resp = httpx.post(
            f'{KOBO2PDF_URL}/render',
            json=body,
            timeout=60,
        )
        if not render_resp.is_success:
            return {'ok': False, 'error': f'kobo2pdf error {render_resp.status_code}: {render_resp.text[:200]}'}

        content_type = render_resp.headers.get('content-type', '')
        if 'application/pdf' not in content_type:
            return {'ok': False, 'error': f'kobo2pdf returned unexpected content-type: {content_type}'}

        return {'ok': True, 'pdf_bytes': render_resp.content}
    except Exception as exc:
        return {'ok': False, 'error': f'Failed to reach kobo2pdf service: {exc}'}
